src/briefcase/platforms/macOS/__init__.py

```diff
@@ -602,26 +602,22 @@ def sign_app(self, app: AppConfig, identity: SigningIdentity):
         sign_targets.append(bundle_path)
 
         # Run signing through a ThreadPoolExecutor so that they run in parallel.
-        # However, we need to ensure that objects are signed from the inside out
-        # (i.e., a folder must be signed *after* all it's contents has been
-        # signed). To do this, we sort the list of signing targets in reverse
-        # lexicographic order, and then group all the signing targets by parent.
-        # This sorts all the signable files into folders; and sign all files in
-        # a folder before sorting the next group. This ensures that longer paths
-        # are signed first, and all files in a folder are signed before the
-        # folder is signed.
+        # However, we need to ensure that objects are signed from the inside out (i.e.,
+        # a folder must be signed *after* all it's contents has been signed, and files
+        # in a folder must be signed *after* all subfolders in that same folder). To do
+        # this, we sort the list of signing targets in depth-first directory order.
         #
-        # NOTE: We are relying on the fact that the final iteration order
-        # produced by groupby() reflects the order in which groups are found in
-        # the input data. The documentation for groupby() says that a new break
-        # is created every time a new group is found in the input data; sorting
-        # the input in reverse order ensures that only one group is found per folder,
-        # and that the deepest folder is found first.
+        # NOTE: We are relying on the fact that the final iteration order produced by
+        # groupby() reflects the order in which groups are found in the input data. The
+        # documentation for groupby() says that a new break is created every time a new
+        # group is found in the input data; sorting the input in reverse order ensures
+        # that only one group is found per folder, and that the deepest folder is found
+        # first.
```

Comment:
But it doesn't guarantee that subfolders and files in the same folder are put into separate groups. For example, if there was no Helpers directory, the group separation would be:
```
.../Qt/lib/QtWebEngineCore.framework/Versions/A/QtWebEngineProcess.app/Contents/MacOS/QtWebEngineProcess
---
.../Qt/lib/QtWebEngineCore.framework/Versions/A/QtWebEngineProcess.app
.../Qt/lib/QtWebEngineCore.framework/Versions/A/QtWebEngineCore
```

Because the last two paths have the same parent, they would be signed in parallel.

Possible solution: make the key for both sorting and grouping be `(path.parent, path.is_dir())`, and sort with `reverse=True`. Those two elements map directly onto the two requirements stated in the comment.

If the grouping was factored out into a utility function, it could be covered by a test.

Reply:
Good catch - I hadn't considered that although the sort order will put the directory first, if they're grouped together parallelism won't guarantee their signed in that order. 

I've modified the grouping as you've described (well - close - the sorting already puts directories before files, so we just need a different grouping key, not a full reverse sort order).

Reply:
WIth `reverse=True`, I think the sorting and grouping key can both be `(path.parent, path.is_dir())`, and all the complexity of `sorted_depth_first` can be removed.

Reply:
The ordering within the group shouldn't matter, but to make the build a bit more reproducible (and pass the tests), it would also be a good idea to sort within each group by name.

Reply:
The key can't be exactly the same, because the sorting process needs to sort the actual files, whereas the grouping needs to sort on the containers of the files. So - we need to have a different key; the three-element key gives the sort order that matches a normal directory listings, so in any debug log it's going to be marginally less confusing. The 3-element key also matches the "complexity" of the actual search - sorting by "depth, then by directory vs file, then by filename" is literally what the algorithm is sorting on, rather than relying on a subtle edge effect of reverse interacting with the boolean value of `is_dir()`.

Reply:
OK, so we can sort the whole list by`key=(path.parent, path.is_dir()), reverse=True`, then group by the same key, then sort each group by `path`. That's still much easier to understand than the current version of `sorted_depth_first`, which creates keys which are *lists* of tuples.

Reply:
Unless the file sorting contains the actual path in the key, the ordering of files/folders inside any given group will be at least partially unpredictable. This won't matter to the signing order, but it makes testing a little harder because we can't *completely* guarantee the order of the output.

It also seems a little silly to me have a `sort` method that doesn't *fully* sort. Plus, we're already sorting, so requiring a *second* pass of sorting (in the test harness, and potentially in the usage of the final groups) seems a little silly.

However, I can't argue that the simple tuple is much easier to understand than the list of tuples, even if the resulting order isn't quite a "lexically pleasing" as the more complex option; and we can fix the filename sorting by adding the actual path to the initial sort key (using the simpler 2 part key for the grouping key).